### PR TITLE
Economy of numbers

### PR DESCRIPTION
**Pull request by Akintomiwaojo**
## Summary/Motivation:
Adds Economy of Number, a scaling function that estimates the future profitability of novel/First-Of-A-Kind (FOAK) equipment.

## Changes proposed in this PR:
-
-

## Reviewer's checklist / merge requirements:
- [ ] The head branch (i.e. the "source" of the changes) is not the `main` branch on the PR author's fork
- [ ] Documentation
- [ ] Tests
- [ ] Diagnostic tests for models

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.md file
   at the top level of this directory.

2. I represent I am authorized to make the contributions and grant the license. If my employer has
   rights to intellectual property that includes these contributions, I represent that I have
   received permission to make contributions and grant the required license on behalf of that
   employer.

### CODE REVIEW

File: src/prommis/uky/costing/ree_plant_capcost.py [hunk 1]
```diff
@@ -3442,3 +3443,70 @@ def assert_Pyomo_object(b, name):
...
+            mutable=True,
+            units=pyunits.dimensionless,
+            doc="The learning factor reflects the level of maturity of the unit/technology",
+        )
+        blk.cost_FOAK = Param(
+            initialize=cost_FOAK,
+            mutable=True,
+            doc="Cost of the First-of-A-Kind of the unit",
+            units=getattr(pyunits, "MUSD_" + CE_index_year),
+        )
```
Comment (bpaul4) [inline src/prommis/uky/costing/ree_plant_capcost.py:3492]: Instead of making this a parameter, why not just use `cost_FOAK` directly? This will also ensure the model is dynamic, and will update/resolve if things change upstream.

File: src/prommis/uky/costing/ree_plant_capcost.py [hunk 2]
```diff
@@ -3442,3 +3443,70 @@ def assert_Pyomo_object(b, name):
...
+        def learning_rate_exponent(b):
+
+            return -log10(1 - b.learning_rate) / log10(2)
+
+        @blk.Constraint()
+        def cost_NOAK_eq(b):
+            return b.cost_NOAK == b.cost_FOAK * (
+                (b.cum_num_units) ** -(b.learning_rate_exponent)
+            )
```
Comment (bpaul4) [inline src/prommis/uky/costing/ree_plant_capcost.py:3512]: Wrap the right hand side in a units conversion, this will catch any case where `cost_FOAK` doesn't have currency units:

```suggestion
        @blk.Constraint()
        def cost_NOAK_eq(b):
            return b.cost_NOAK == pyunits.convert(
                b.cost_FOAK * (
                    (b.cum_num_units) ** -(b.learning_rate_exponent)
                ),
                to_units=getattr(pyunits, "MUSD_" + CE_index_year)
            )
```

File: src/prommis/uky/costing/tests/test_ree_costing.py [hunk 3]
```diff
@@ -6314,3 +6314,32 @@ def test_REE_costing_royalty_expression_solve():
...
+
+    dt = DiagnosticsToolbox(model=m, variable_bounds_violation_tolerance=1e-4)
+    dt.assert_no_structural_warnings()
+
+    solver = get_solver()
+    results = solver.solve(m, tee=True)
+    assert_optimal_termination(results)
+    dt.assert_no_numerical_warnings()
+
+    assert pyo.value(m.fs.costing.cost_NOAK[None]) == pytest.approx(36.574, rel=1e-4)
```
Comment (bpaul4) [inline src/prommis/uky/costing/tests/test_ree_costing.py:6345]: Why is `cost_NOAK` indexed here?

Comment (Akintomiwaojo) [reply src/prommis/uky/costing/tests/test_ree_costing.py:6345]: I used `None` similarly to how it was used for pv_royalties, like accessing a scalar value of an expression. But looking closer now, I can see I don't need to index `cost_NOAK` as it is a variable, not an expression. Thanks for pointing this out.

File: src/prommis/uky/costing/ree_plant_capcost.py [hunk 4]
```diff
@@ -3442,3 +3443,65 @@ def assert_Pyomo_object(b, name):
                     f"Argument {obj} of type {type(obj)} is not a supported object type. "
                     f"Ensure {name} is a Pyomo Param, Var, Expression, or ScalarExpression."
                 )
+
+    def economy_of_numbers(
+        blk, cum_num_units, cost_FOAK, CE_index_year, learning_rate=0.04
+    ):
+        """
+        Equations for Economy of Numbers (EoN) derived from NETL Quality Guidelines for Energy System Studies
```
Comment (Agfritz) [inline src/prommis/uky/costing/ree_plant_capcost.py:3451]: It looks like this QGESS is no longer publicly released (but i do see the references to it in other documents). Were you able to find a version of it?

File: src/prommis/uky/costing/ree_plant_capcost.py [hunk 5]
```diff
@@ -3442,3 +3443,65 @@ def assert_Pyomo_object(b, name):
...
+        b = - log(1-R)/log(2)
+        where Y is the cost of the Nth-of-A-Kind (NOAK) of the equipment, A is the cost of the FOAK,
+        X is the cumulative number of units, b is the learning rate exponent, and R is the learning
+        rate constant.
+
+        Args:
+            cum_num_units: The cumulative number of units.
+            cost_FOAK: The cost of manufacturing the First-of-A-Kind equipment.
+            CE_index_year: year for cost basis, e.g., "2021" to use 2021 dollars
+            learning_rate: ranges between 0.01 - 0.06, depending on the level of maturity
```
[the remark this answers is not in the document]
Comment (Agfritz) [reply src/prommis/uky/costing/ree_plant_capcost.py:3470]: In a review of the QGESS, it shows a range of 0.01-0.1. https://www.cmu.edu/epp/iecm/rubin/PDF%20files/2016/An%20Assessment%20of%20NETL%20Cost%20Estimation%20Methodology_20160429.pdf

would it make sense to increase the range and to provide a link to the table recommending learning rate by maturity level?

File: src/prommis/uky/costing/ree_plant_capcost.py (hunk 4 above)
Comment (Akintomiwaojo) [reply src/prommis/uky/costing/ree_plant_capcost.py:3451]: I just realized my version is missing from the official NETL page. However, the info is also in the review of the QGESS you shared earlier. Should I cite the review document or the QGESS report, considering that the review also cited the QGESS report?

Comment (bpaul4) [reply src/prommis/uky/costing/ree_plant_capcost.py:3451]: See my comment above.

If you cite the QGESS, check that your version is publicly available or reference a newer version with the same information. Your version might be on OSTI or another database.

Comment (Akintomiwaojo) [reply src/prommis/uky/costing/ree_plant_capcost.py:3451]: Unfortunately, it's also not on OSTI or any other database. Since it has the same information, I will use the CMU report. Thank you.

Comment (Agfritz) [reply src/prommis/uky/costing/ree_plant_capcost.py:3451]: Ok thank you! It looks like the reference is fixed below but I would also recommend removing the reference to QGESS here since it is no longer published. I also did not see that the CMU report published the equations, were they in the report or did i miss them?

Comment (Akintomiwaojo) [reply src/prommis/uky/costing/ree_plant_capcost.py:3451]: The CMU report also published the equations (Chapter 6, equations 2 and 3). However, the definition of variables in the report is slightly different from what I used in the code. For instance, they worked with a reduction in time as the cumulative production unit increased, and for ours, it's cost reduction.

Comment (Agfritz) [reply src/prommis/uky/costing/ree_plant_capcost.py]: The equation formation used is shown in this publication: https://www.osti.gov/pages/biblio/1868500
I would recommend adding this as the citation for the equations.

**Review by bpaul4 — APPROVED** [review]
I think this looks great! Thanks @Akintomiwaojo for contributing this.

**Review by Agfritz — APPROVED** [review]
Thanks for making all the changes so quickly Tomi, looks good to go!
